[Bug] Error opening collectin on Android 12 Beta

###### Reproduction Steps

1. Install Android 12 beta
2. Open Ankidroid
3. See error opening collection.


###### Expected Result
App opens



###### Actual Result
App doesnt open, cites an error opening collection



###### Debug info

Here's the logcat output:
	```
12:50:09.765	AnkiDroid		CollectionHelper/ Failed to query version
com.ichi2.libanki.exception.UnknownDatabaseVersionException: java.lang.NullPointerException: Attempt to invoke virtual method 'int com.ichi2.libanki.DB.queryScalar(java.lang.String, java.lang.Object[])' on a null object reference
	at com.ichi2.libanki.Collection.queryVer(Collection.java:2)
	at com.ichi2.anki.CollectionHelper.getDatabaseVersion(CollectionHelper.java:2)
	at com.ichi2.anki.CollectionHelper.isFutureAnkiDroidVersion(CollectionHelper.java:1)
	at com.ichi2.anki.InitialActivity.isFutureAnkiDroidVersion(InitialActivity.java:1)
	at com.ichi2.anki.InitialActivity.getStartupFailureType(InitialActivity.java:5)
	at com.ichi2.anki.DeckPicker.handleStartup(DeckPicker.java:5)
	at com.ichi2.anki.DeckPicker.onCreate(DeckPicker.java:6)
	at android.app.Activity.performCreate(Activity.java:8036)
	at android.app.Activity.performCreate(Activity.java:8016)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3581)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3765)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:102)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2193)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7796)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:974)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'int com.ichi2.libanki.DB.queryScalar(java.lang.String, java.lang.Object[])' on a null object reference
	at com.ichi2.libanki.Collection.queryVer(Collection.java:1)
	... 22 more

com.ichi2.libanki.exception.UnknownDatabaseVersionException: java.lang.NullPointerException: Attempt to invoke virtual method 'int com.ichi2.libanki.DB.queryScalar(java.lang.String, java.lang.Object[])' on a null object reference
	at com.ichi2.libanki.Collection.queryVer(Collection.java:2)
	at com.ichi2.anki.CollectionHelper.getDatabaseVersion(CollectionHelper.java:2)
	at com.ichi2.anki.CollectionHelper.isFutureAnkiDroidVersion(CollectionHelper.java:1)
	at com.ichi2.anki.InitialActivity.isFutureAnkiDroidVersion(InitialActivity.java:1)
	at com.ichi2.anki.InitialActivity.getStartupFailureType(InitialActivity.java:5)
	at com.ichi2.anki.DeckPicker.handleStartup(DeckPicker.java:5)
	at com.ichi2.anki.DeckPicker.onCreate(DeckPicker.java:6)
	at android.app.Activity.performCreate(Activity.java:8036)
	at android.app.Activity.performCreate(Activity.java:8016)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3581)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3765)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:102)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2193)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7796)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
```
Refer to the [support page](https://ankidroid.org/docs/help.html) if you are unsure where to get the "debug info".

###### Research
*Enter an [x] character to confirm the points below:*

- [x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] I have searched for similar existing issues here and on the user forum
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

--- a/AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/libanki/DBTest.java
+++ b/AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/libanki/DBTest.java
@@ -1,11 +1,14 @@
 package com.ichi2.anki.tests.libanki;
 
 import android.Manifest;
+import android.app.Application;
+import android.content.Context;
 import android.database.sqlite.SQLiteDatabase;
 import android.database.sqlite.SQLiteDatabaseCorruptException;
 import android.os.Build;
 
 import com.ichi2.anki.CollectionHelper;
+import com.ichi2.anki.tests.InstrumentedTest;
 import com.ichi2.compat.CompatHelper;
 import com.ichi2.libanki.DB;
 
@@ -19,11 +22,13 @@ import java.io.FileOutputStream;
 import java.util.Random;
 
 import androidx.sqlite.db.SupportSQLiteDatabase;
+import androidx.test.core.app.ApplicationProvider;
 import androidx.test.ext.junit.runners.AndroidJUnit4;
+import androidx.test.platform.app.InstrumentationRegistry;
 import androidx.test.rule.GrantPermissionRule;
 
 @RunWith(AndroidJUnit4.class)
-public class DBTest {
+public class DBTest extends InstrumentedTest {
 
     @Rule
     public GrantPermissionRule mRuntimePermissionRule =
@@ -32,7 +37,7 @@ public class DBTest {
     @Test
     public void testDBCorruption() throws Exception {
 
-        String storagePath = CollectionHelper.getDefaultAnkiDroidDirectory();
+        String storagePath = CollectionHelper.getDefaultAnkiDroidDirectory(getTestContext());
         File illFatedDBFile = new File(storagePath, "illFatedDB.anki2");
 
         // Make sure we have clean state to start with

--- a/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java
@@ -348,7 +348,7 @@ public class AnkiDroidApp extends Application {
                 CollectionHelper.initializeAnkiDroidDirectory(dir);
             } catch (StorageAccessException e) {
                 Timber.e(e, "Could not initialize AnkiDroid directory");
-                String defaultDir = CollectionHelper.getDefaultAnkiDroidDirectory();
+                String defaultDir = CollectionHelper.getDefaultAnkiDroidDirectory(this);
                 if (isSdCardMounted() && CollectionHelper.getCurrentAnkiDroidDirectory(this).equals(defaultDir)) {
                     // Don't send report if the user is using a custom directory as SD cards trip up here a lot
                     sendExceptionReport(e, "AnkiDroidApp.onCreate");

--- a/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java
@@ -151,6 +151,17 @@ import static timber.log.Timber.DebugTree;
 )
 public class AnkiDroidApp extends Application {
 
+    /**
+     * Toggles Scoped Storage functionality introduced in later commits <p>
+     * Can be set to true or false only by altering the declaration itself.
+     * This restriction ensures that this flag will only be used by developers for testing <p>
+     * Set to false by default, so won't migrate data or use new scoped dirs <p>
+     * If true, enables data migration & use of scoped dirs in later commits <p>
+     * Should be set to true for testing Scoped Storage <p>
+     * TODO: Should be removed once app is fully functional under Scoped Storage
+     */
+    public static final boolean TESTING_SCOPED_STORAGE = false;
+
     private static final String WEBVIEW_VER_NAME = "WEBVIEW_VER_NAME";
 
     public static final String XML_CUSTOM_NAMESPACE = "http://arbitrary.app.namespace/com.ichi2.anki";

--- a/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java
@@ -237,14 +237,115 @@ public class CollectionHelper {
 
     /**
      * Get the absolute path to a directory that is suitable to be the default starting location
-     * for the AnkiDroid folder. This is a folder named "AnkiDroid" at the top level of the
-     * external storage directory.
-     * @return the folder path
+     * for the AnkiDroid folder.
+     * <p>
+     * Currently, this is a folder named "AnkiDroid" at the top level of the non-app-specific external storage directory.
+     * <p><br>
+     * When targeting API > 29, AnkiDroid will have to use Scoped Storage on any device of any API level.
+     * Scoped Storage only allows access to App-Specific directories (without permissions).
+     * Hence, AnkiDroid won't be able to access the directory used currently on all devices,
+     * regardless of their API level, once AnkiDroid targets API > 29.
+     * Instead, AnkiDroid will have to use an App-Specific directory to store the AnkiDroid folder.
+     * This applies to the entire AnkiDroid userbase.
+     * <p><br>
+     * Currently, if <code>TESTING_SCOPED_STORAGE</code> is set to <code>true</code>, AnkiDroid uses its External
+     * App-Specific directory.<p>
+     * External App-Specific Storage is used since the only advantage Internal App-Specific Storage has over External
+     * App-Specific storage is additional security, but AnkiDroid does not store sensitive data. Defaulting to
+     * External Storage preserves the current behavior of the App
+     * (AnkiDroid defaults to External before the Migration To Scoped Storage).
+     * <p>
+     * TODO: If External Storage isn't emulated, allow users to choose between External & Internal App-Specific Storage
+     *  instead of defaulting to External App-Specific Storage. This should be done since using either one may be more
+     *  useful for them. If External Storage is emulated, there is no use in providing the option since Internal
+     *  Storage can not provide more storage space than External Storage if External Storage is emulated.
+     * <p><br>
+     * See the detailed explanation on storage locations & their classification below for more details.
+     * <p><br>
+     * App-Specific storage refers to directories which are meant to store files that are meant to be used by a
+     * particular app. Each app has its own Internal & External App-Specific directory. Under Scoped Storage,
+     * an app can only access its own Internal & External App-Specific directory without needing permissions.
+     * <p><br>
+     * Storage can be classified as Internal or External Storage. <p><br>
+     * Internal Storage: This storage is characterized by the fact that it is always available since it always resides
+     * on the device's own non-removable storage.<p>
+     * App-Specific Internal Storage can be accessed by ONLY the app which owns that directory (without any permissions).
+     * It cannot be accessed by any other apps.
+     * It cannot be accessed using the Files app on Android or by connecting a device to a pc via USB.
+     * <p><br>
+     * External Storage: <p>
+     * This storage is characterized only by the fact that it is not guaranteed to be available.<p>
+     * It may be built-in, non-removable storage on the device which is being emulated to function like external storage.
+     * In this case, it doesn't offer more space than Internal Storage.<p>
+     * Or, it may be removable storage like an SD Card.<p>
+     * App-Specific External Storage can be accessed by the app it is owned by without any permissions.
+     * It can be accessed by any apps with the WRITE_EXTERNAL_STORAGE permission.
+     * It can also be accessed via the Android Files app or by connecting the device to a PC via USB.
+     * <p><br>
+     * Note: The Files app can be misleading. On Samsung devices, clicking on Internal Storage it actually shows the
+     * emulated external storage (/storage/emulated/0/ in my case) - this is because from the point of view of the user,
+     * emulated external storage is just more internal storage since it is built into the phone. This is why vendors
+     * like Samsung may refer to external emulated storage as internal storage, even though for developers, they mean
+     * very different things as explained above.
+     * <p><br>
+     *
+     * @return Absolute Path to the default location starting location for the AnkiDroid folder
      */
     @SuppressWarnings("deprecation") // TODO Tracked in https://github.com/ankidroid/Anki-Android/issues/5304
     @CheckResult
     public static String getDefaultAnkiDroidDirectory(@NonNull Context context) {
+        if (AnkiDroidApp.TESTING_SCOPED_STORAGE) {
+            return getAppSpecificExternalAnkiDroidDirectory(context);
+        }
+        return getLegacyAnkiDroidDirectory();
+    }
+
+
+    /**
+     * Returns the absolute path to the AnkiDroid directory under the primary/shared external storage directory.
+     * This directory may be in emulated external storage, or can be an SD Card directory.
+     * <p>
+     * The path returned will no longer be accessible to AnkiDroid once targetSdk > 29
+     *
+     * @return Absolute path to the AnkiDroid directory in primary shared/external storage
+     */
+    @SuppressWarnings("deprecation")
+    public static String getLegacyAnkiDroidDirectory() {
         return new File(Environment.getExternalStorageDirectory(), "AnkiDroid").getAbsolutePath();
+    }
+
+
+    /**
+     * Returns the absolute path to the AnkiDroid directory under the app-specific, primary/shared external storage
+     * directory.
+     * <p>
+     * This directory may be in emulated external storage, or can be an SD Card directory.
+     * If it is actually external storage, i.e., removable storage like an SD Card, instead of storage
+     * built into the device itself, using this directory over internal storage can be beneficial since
+     * it may be able to store more data.
+     * <p>
+     * AnkiDroid can access this directory without permissions, even under Scoped Storage
+     * Other apps can access this directory if they have the WRITE_EXTERNAL_STORAGE permission
+     *
+     * @param context Used to get the External App-Specific directory for AnkiDroid
+     * @return Returns the absolute path to the App-Specific External AnkiDroid directory
+     */
+    public static String getAppSpecificExternalAnkiDroidDirectory(@NonNull Context context) {
+        return context.getExternalFilesDir(null).getAbsolutePath();
+    }
+
+
+    /**
+     * Returns the absolute path to the private AnkiDroid directory under the app-specific, internal storage directory.
+     * <p>
+     * AnkiDroid can access this directory without permissions, even under Scoped Storage
+     * Other apps cannot access this directory, regardless of what permissions they have
+     *
+     * @param context Used to get the Internal App-Specific directory for AnkiDroid
+     * @return Returns the absolute path to the App-Specific Internal AnkiDroid Directory
+     */
+    public static String getAppSpecificInternalAnkiDroidDirectory(@NonNull Context context) {
+        return context.getFilesDir().getAbsolutePath();
     }
 
     /**

--- a/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java
@@ -243,7 +243,7 @@ public class CollectionHelper {
      */
     @SuppressWarnings("deprecation") // TODO Tracked in https://github.com/ankidroid/Anki-Android/issues/5304
     @CheckResult
-    public static String getDefaultAnkiDroidDirectory() {
+    public static String getDefaultAnkiDroidDirectory(@NonNull Context context) {
         return new File(Environment.getExternalStorageDirectory(), "AnkiDroid").getAbsolutePath();
     }
 
@@ -264,7 +264,7 @@ public class CollectionHelper {
         return PreferenceExtensions.getOrSetString(
                 preferences,
                 "deckPath",
-                CollectionHelper::getDefaultAnkiDroidDirectory);
+                () -> getDefaultAnkiDroidDirectory(context));
     }
 
     /**

--- a/AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java
@@ -349,7 +349,7 @@ public class Preferences extends AppCompatPreferenceActivity implements Preferen
                                 .positiveText(R.string.dialog_ok)
                                 .negativeText(R.string.reset_custom_buttons)
                                 .onPositive((dialog, which) -> dialog.dismiss())
-                                .onNegative((dialog, which) -> collectionPathPreference.setText(CollectionHelper.getDefaultAnkiDroidDirectory()))
+                                .onNegative((dialog, which) -> collectionPathPreference.setText(CollectionHelper.getDefaultAnkiDroidDirectory(this)))
                                 .show();
                         return false;
                     }

